Documentation/RepoApi.md

```diff
@@ -0,0 +1,137 @@
+# Standard Repo API
+
+In order to build the .NET Core product from its constiuent repos, there needs to be a standardized set of actions
+each repo can perform. Not only will this make building the repos easier, it will make it easier to onboard
+new repos. If a new repo conforms to the standardized set of actions, source-build can easily start consuming it.
+
+This document describes the basic structure of this "Repo API", and the requirements imposed on a repo that needs
+to participate in source-build.
+
+## build.cmd/sh
+
+Each repo will have two files that implement the repo API:
+
+* build.cmd
+* build.sh
+
+The `.cmd` file will be called when running on Windows, while the `.sh` file will be called on all other operating systems.
+
+These scripts will contain the logic necessary in order to build, test, publish the repo's assets.
+
+source-build will invoke these scripts on each repo passing in:
+
+1. An action to perform
+1. A set of parameters
+
+Individual repos are free to implement more features in these build scripts than what is required by source-build. For
+example, a repo may want to provide a default action. These extra behaviors are not defined by source-build.
+
+### Actions
+
+To start, there are a number of actions source-build will need the repos to implement.  The actions are:
+
+1. `Restore` - Sync any external dependencies/tools necessary to build the repo.
+1. `Build` - Assemble the product from the current source contained in the repo.
+1. `Publish` - Put the assembled product in the specified location.
+1. `Test` - Ensure the assembled product works/acts correctly.
+
+This list of actions can be added to in the future as more scenarios are implemented.
+
+source-build formats arguments using a style that is compatible with MSBuild. The reasoning for this is that the
+majority of repos have an MSBuild based build system. Using this format will allow for arguments to easily be passed
+through to MSBuild. Similarly, both action and parameter arguments are case-insensitive, just like MSBuild default
+behavior.
+
+Actions will take the form of `-t:ActionName`, just like MSBuild targets are passed to `msbuild.exe`.
+
+### Parameters
+
+Each action above will take a set of parameters used to pass information into the action and control its behavior.
+
+Parameters will take the form of `-p:ParameterName=ParameterValue`, just like MSBuild properties are passed to `msbuild.exe`.
+One slight variation here is the usage of `-p` instead of `/p`.  MSBuild supports both forms, but source-build chooses `-p` because
```

Comment:
FYI: msbuild that runs on .NET framework doesn't support "-p".

Reply:
Yes it does.

Reply:
You are correct they do support it. For some reason they aren't in my PS wrapper. This is what I was getting:

```
> msbv -p:P=v
MSBUILD : error MSB1005: Specify a property and its value.
Switch: -p:

For switch syntax, type "MSBuild /help"
```

Reply:
what version of msbuild is that?

Reply:
15.3.409.57025

Please note that if I call `msbuild.exe -p:P=v` directly it accepts it but my powershell wrapper function for some reason doesn't work. In case you are curious the wrapper function is just this `function msbv { msbuild /nologo /flp:verbosity=diag $args }`

Reply:
Powershell functions accept `-p:P=v` as setting parameter `p` to `P=v`, and sets `$args` to an array with `-p:` and `P=v`. That gives `msbuild` `-p: Foo=Bar`. (Annoying) workaround is to call `msbv '-p:P=v'`.